### ISSUE TITLE
Accuracy calculation is incorrect?

### ISSUE THREAD

**Issue by Horsmann** [fchollet/keras#4336]
Hi,
I have been experimenting with using `model.predict()` to run some predictions manually rather than evaluating them directly with `model.evaluate()` because I need the predicted values and not just the overall accuracy.

When computing the accuracy manually I noticed that the accuracy keras' calculates in `evaluate()` is not correct. Is there a bug in the accuracy calculation?

Below a self-contained runnable minimal example that shows the difference (the example tests on the training data):
```
import numpy as np
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Activation, Embedding, TimeDistributed, Bidirectional
from keras.layers import LSTM
import json
from keras.utils import np_utils

np.set_printoptions(threshold=np.nan)

def mapWords2Integer(input, startIdx=0):
    map = {}
    for s in input:
        for w in s:
            if w not in map:
               map[w]=startIdx
               startIdx+=1

    out = []
    for s in input:
        out_s = []
        for w in s:
            out_s.append(map[w])
        out.append(out_s)
    return map, out

tokens = [
    ['Great', 'Western', 'said', 'it', 'had', 'a', 'sharp', 'increase', 'in', 'margins', 'in', 'the', 'recent', 'third', 'quarter', '.'],
    ['Margins', 'are', 'the', 'difference', 'between', 'the', 'yield', 'on', 'the', 'company', 's', 'earning', 'assets', 'and', 'its', 'own', 'cost', 'of', 'funds', '.'],
    ['But', 'a', 'reduction', 'in', 'one-time', 'gains', 'on', 'the', 'sale', 'of', 'various', 'assets', 'and', 'an', 'increase', 'in', 'the', 'company', 's', 'provision', 'for ', 'loan', 'losses', 'held', 'down', 'the', 'earnings', 'gain', ', ', 'the', 'company', 'said', '.']
]

labels = [
    ['NNP', 'NNP', 'VBD', 'PRP', 'VBD', 'DT', 'JJ', 'NN', 'IN', 'NNS', 'IN', 'DT', 'JJ', 'JJ', 'NN', '.'],
    ['NNS', 'VBP', 'DT', 'NN', 'IN', 'DT', 'NN', 'IN', 'DT', 'NN', 'POS', 'VBG', 'NNS', 'CC', 'PRP$', 'JJ', 'NN', 'IN', 'NNS', '.'],
    ['CC', 'DT', 'NN', 'IN', 'JJ', 'NNS', 'IN', 'DT', 'NN', 'IN', 'JJ', 'NNS', 'CC', 'DT', 'NN', 'IN', 'DT', 'NN', 'POS', 'NN', 'IN', 'NN', 'NNS', 'VBD', 'RP', 'DT', 'NNS', 'NN', ',', 'DT', 'NN', 'VBD', '.']
]


longest_sequence = max(len(s) for s in (tokens))

valMap, valInt = mapWords2Integer(tokens,1)
labMap, labelInt = mapWords2Integer(labels,1)

vocabSize=len(valMap)

padValInt = sequence.pad_sequences(valInt, maxlen=longest_sequence)
padLabInt = sequence.pad_sequences(labelInt, maxlen=longest_sequence)

maximal_value = max([ys for sent in padLabInt for ys in sent])+1

train_label = np.array([np_utils.to_categorical(seq, maximal_value) for seq in padLabInt])

EMBEDDING_DIM=100

model = Sequential()
model.add(Embedding(vocabSize+1, EMBEDDING_DIM,mask_zero=True))
model.add(Bidirectional(LSTM(EMBEDDING_DIM, return_sequences=True)))
model.add(TimeDistributed(Dense(maximal_value)))
model.add(Activation('relu'))

# try using different optimizers and different optimizer configs
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model.fit(padValInt, train_label, nb_epoch=1)
score, acc = model.evaluate(padValInt, train_label)
print('Accuracy calculated by Keras:', acc*100)

#########################
##### Manual Accuracy ###
#########################
correct=0.0
incorrect=0.0
inversed_label_map = {v: k for k, v in labMap.items()}
for k in range(0, len(padValInt)):
    s = padValInt[k]
    out = model.predict(s)
    predLabels = []
    for e in out:
        for c in e:
            x = np.argmax(c)
            pl = inversed_label_map.get(x)
            predLabels.append(pl)

    #throw away the padded, leading zeros
    fromIdx = (len(predLabels)-len(labels[k]))
    predLabels = predLabels[fromIdx:]

    for i in range(0, len(labels[k])):

        if labels[k][i] == predLabels[i]:
            correct+=1
        else:
            incorrect+=1

acc = correct / (correct+incorrect) * 100
print("Manually calculated accuracy: ", acc)
```

**Comment by pechersky** [comment]
Without running your code, a possibility is that the model is running a different type of accuracy metric than you expect. Given your `predLabels` and calculated accuracy, what do you get when you run the following?
```
keras.metrics.binary_accuracy(labels, predLabels)
keras.metrics.categorical_accuracy(labels, predLabels)
keras.metrics.sparse_categorical_accuracy(labels, predLabels)
```

**Comment by Horsmann** [comment]
Hi,
I get different accuracy values for the metrics you mention but non is the one that I calculate by hand. 
The difference is sometimes only small but I have cases in which Keras says the accuracy is `zero` but when printing gold/prediction values I get at least one or two predictions right.

**Comment by phdowling** [comment]
I think Keras does some rounding internally, could that be it? How different are your values?

**Comment by Horsmann** [comment]
I modified my example from the first post with the new initialisation of the variables. I still have differences which cannot be explained by rounding. Here the results I get after running 4x times above code:

```
Accuracy calculated by Keras: 5.05050505051
Manually calculated accuracy:  7.246376811594203

Accuracy calculated by Keras: 14.1414141414
Manually calculated accuracy:  14.492753623188406

Accuracy calculated by Keras: 18.1818181818
Manually calculated accuracy:  21.73913043478261

Accuracy calculated by Keras: 7.07070707071
Manually calculated accuracy:  5.797101449275362
```

Either I am doing something fundamentally wrong with retrieving/converting the output of Keras or the the accuracy Keras computes is wrong

**Comment by janiteja** [comment]
@Horsmann Yes, Even I had this problem. I don't know what 'accuracy' metric is calculating but it is far beyond the actual accuracy. Try to give 'categorical_accuracy' as your metric. That will give you correct values.

**Comment by cbaziotis** [comment]
I had the same problem with wrong scores and i "solved" it by using the sklearn's metrics in a custom callback.

**Comment by lebavarois** [comment]
@Horsmann When I try to run the code example you provided, I get the following error:

> Exception: A target array with shape (3, 33, 17) was passed for an output of shape (None, None, 17) while using as loss `categorical_crossentropy`. This loss expects targets to have the same shape as the output.

**Comment by Horsmann** [comment]
@lebavarois I use the Theano backend. I just tried the code again it (still) works for me

My problem was the zero padding. I kind of expected that Keras would exclude the zero-padded values from its evaluation but it doesn't.
